src/chainstate/stacks/index/trie.rs

```diff
@@ -0,0 +1,2053 @@
+        assert!(cursor.node_ptrs.len() > 0);
+
+        let mut ptrs = cursor.node_ptrs.clone();
+        trace!("update_root_hash: ptrs = {:?}", &ptrs);
+        let mut child_ptr = ptrs.pop().unwrap();
+
+        if ptrs.len() == 0 {
+            // root node was already updated by trie operations, but it will have the wrong hash.
+            // we need to "fix" the root node so it mixes in its ancestor hashes.
+            trace!("Fix up root node so it mixes in its ancestor hashes");
+            let (node, cur_hash) = storage.read_nodetype(&child_ptr)?;
+            if node.is_node256() {
+                let root_disk_ptr = storage.root_ptr();
+                let root_ptr = TriePtr::new(TrieNodeID::Node256, 0, root_disk_ptr as u32);
+                if child_ptr != root_ptr {
+                    return Err(Error::CorruptionError("Only ptr is not the root".to_string()));
+                }
+                
+                let mut hash_buf = Vec::with_capacity(TRIEHASH_ENCODED_SIZE * 256);
+                Trie::get_children_hashes_bytes(storage, &node, &mut hash_buf)?;
+
+                let h = 
+                    if update_skiplist {
+                        trace!("Update root skiplist");
+                        Trie::get_trie_root_hash(storage, &get_nodetype_hash_bytes(&node, &hash_buf))?
+                    }
+                    else {
+                        trace!("Not updating root skiplist");
+                        get_nodetype_hash_bytes(&node, &hash_buf)
+                    };
+
+                // for debug purposes
+                if is_trace() {
+                    let node_hash = get_nodetype_hash_bytes(&node, &hash_buf);
+                    let hs = Trie::get_trie_root_ancestor_hashes(storage, &node_hash)?;
+                    trace!("update_root_hash: Updated {:?} with {:?} from {:?} to {:?} + {:?} = {:?} (fixed root)", &node, &child_ptr, &cur_hash, &node_hash, &hs[1..].to_vec(), &h);
+                }
+
+                storage.write_nodetype(child_ptr.ptr(), &node, h)?;
+            }
+            else {
+                return Err(Error::CorruptionError("Only ptr was not a node256".to_string()));
+            }
+        }
+        else {
+            while ptrs.len() > 0 {
+                let ptr = match ptrs.pop() {
+                    Some(p) => p,
+                    None => panic!("Out of ptrs")
+                };
+                if is_backptr(ptr.id()) {
+                    // this node was not altered, but instead queued to the cursor as part of walking a
+                    // backptr skiplist.  Do nothing.
+                    continue;
+                }
+
+                let (mut node, cur_hash) = storage.read_nodetype(&ptr)?;
+                assert!(!node.is_leaf());
+
+                // this child_ptr _must_ be in the node.
+                let updated = node.replace(&child_ptr);
```

Comment:
I can't follow why `node` should have a child replaced here -- if a child pointer in the path changed, isn't its parent updated at the time of the change?

Reply:
The `update_root_hash()` method _should not_ insert any new nodes.  Insertion is handled by one of the methods reachable from `add_value()`.  The `.replace()` method fails if the given ptr slot isn't filled in the given node.

Reply:
I see -- `.replace` is being used to test whether or not the node contains `child_ptr`